### PR TITLE
Migrate AT_DISPATCH_COMPLEX_TYPES to c10::complex

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This brings the whole `AT_DISPATCH_COMPLEX_TYPES` to `c10::complex`. There are many unnecessary
```C++
if (complex) {
   AT_DISPATCH_COMPLEX_TYPES
} else {
  AT_DISPATCH_...
}
```
logic, and I merged these cases when possible.

CPU kernels are fine, but some CUDA kernels are doing `c10::complex -> thrust::complex` casting, which will be removed later.

### CODE REVIEW

**Review by anjali411 — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
The MacOS failure looks real

File: c10/util/complex_type.h [hunk 1]
```diff
@@ -198,7 +198,7 @@ struct alignas(sizeof(T) * 2) complex_common {
   }
 
   template<typename U>
-  constexpr complex<T> &operator /=(const complex<U> &rhs) {
+  constexpr complex<T> &operator /=(const complex<U> &rhs) __ubsan_ignore_float_divide_by_zero__ {
```
Comment (anjali411) [inline c10/util/complex_type.h:201]: cc. @mruberry for the sanitization check and the failing macos test

Comment (mruberry) [reply c10/util/complex_type.h:201]: ubsan line LGTM; I'm not sure what's going on with the failing Mac test but I agree the issue looks legitimate. I think you'll have to debug it.

**Comment by zasdfgbnm** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/39043 where `ci-all` is enabled so that I can see the result of no-AVX test.

**Pull request closed without merging**